src/NUnitFramework/framework/Assert.cs

```diff
@@ -270,16 +272,95 @@ public static void Contains(object expected, ICollection actual)
 
         #region Multiple
 
-        ///// <summary>
-        ///// If an assert fails within this block, execution will continue and 
-        ///// the errors will be reported at the end of the block.
-        ///// </summary>
-        ///// <param name="del">The test delegate</param>
-        //public static void Multiple(TestDelegate del)
-        //{
-        //    del();
-        //}
+        /// <summary>
+        /// Wraps code containing a series of assertions, which should all
+        /// be executed, even if they fail. Failed results are saved and
+        /// reported at the end of the code block.
+        /// </summary>
+        /// <param name="testDelegate">A TestDelegate to be executed in Multiple Assertion mode.</param>
+        public static void Multiple(TestDelegate testDelegate)
+        {
+            TestExecutionContext context = TestExecutionContext.CurrentContext;
+            Guard.OperationValid(context != null, "Assert.Multiple called outside of a valid TestExecutionContext");
+
+            context.MultipleAssertLevel++;
+
+            try
+            {
+                testDelegate();
+            }
+            finally
+            {
+                context.MultipleAssertLevel--;
+            }
+
+            if (context.MultipleAssertLevel == 0)
+            {
+                int count = context.CurrentResult.AssertionResults.Count;
+
+                if (count > 0)
+                {
+                    var writer = new TextMessageWriter("Multiple Assert block had {0} failure(s).", count);
+                    writer.WriteLine();
+
+                    int counter = 0;
+                    foreach (var assertion in context.CurrentResult.AssertionResults)
+                    {
+                        writer.WriteLine(string.Format("  {0}) {1}", ++counter, assertion.Message));
+                        writer.WriteLine();
+                    }
+
+                    writer.Write(' '); // So the last newline doesn't get dropped
+
+                    throw new AssertionException(writer.ToString());
+                }
+            }
+        }
 
         #endregion
+
+        #region Helper Methods
+
+        private static void ReportFailure(ConstraintResult result, string message)
+        {
+            ReportFailure(result, message, null);
+        }
+
+        private static void ReportFailure(ConstraintResult result, string message, params object[] args)
+        {
+            MessageWriter writer = new TextMessageWriter(message, args);
+            result.WriteMessageTo(writer);
+            string formattedMessage = writer.ToString();
+            string stackTrace = null;
+
+#if PORTABLE
+            try
+            {
+                // Throw to get stack trace for recording the assertion
+                throw new Exception();
+            }
+            catch (Exception ex)
+            {
+                stackTrace = ex.StackTrace;
```

Comment:
I think this is recording way too much stack information. For the `MultipleAssertFailureDemo` method, it records:

```
   at System.Environment.GetStackTrace(Exception e, Boolean needFileInfo)
   at System.Environment.get_StackTrace()
   at NUnit.Framework.Assert.ReportFailure(ConstraintResult result, String message, Object[] args) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Assert.cs:line 348
   at NUnit.Framework.Assert.That[TActual](TActual actual, IResolveConstraint expression, String message, Object[] args) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Assert.That.cs:line 247
   at NUnit.Framework.Assertions.Tests.AssertMultipleTests.<>c.<MultipleAssertFailureDemo>b__3_0() in C:\Source\GitHub\nunit\src\NUnitFramework\tests\Assertions\AssertMultipleTests.cs:line 100
   at NUnit.Framework.Assert.Multiple(TestDelegate testDelegate) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Assert.cs:line 290
   at NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo() in C:\Source\GitHub\nunit\src\NUnitFramework\tests\Assertions\AssertMultipleTests.cs:line 97
   at System.RuntimeMethodHandle._InvokeMethodFast(Object target, Object[] arguments, SignatureStruct& sig, MethodAttributes methodAttributes, RuntimeTypeHandle typeOwner)
   at System.RuntimeMethodHandle.InvokeMethodFast(Object target, Object[] arguments, Signature sig, MethodAttributes methodAttributes, RuntimeTypeHandle typeOwner)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture, Boolean skipVisibilityChecks)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at NUnit.Framework.Internal.Reflect.InvokeMethod(MethodInfo method, Object fixture, Object[] args) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Reflect.cs:line 206
   at NUnit.Framework.Internal.MethodWrapper.Invoke(Object fixture, Object[] args) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\MethodWrapper.cs:line 192
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunNonAsyncTestMethod(TestExecutionContext context) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 103
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(TestExecutionContext context) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 78
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 60
   at NUnit.Framework.Internal.Commands.TestActionCommand.Execute(TestExecutionContext context) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Commands\TestActionCommand.cs:line 81
   at NUnit.Framework.Internal.Commands.SetUpTearDownCommand.Execute(TestExecutionContext context) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Commands\SetUpTearDownCommand.cs:line 66
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItem.cs:line 59
   at NUnit.Framework.Internal.Execution.WorkItem.RunTest() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 360
   at NUnit.Framework.Internal.Execution.WorkItem.Execute() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 260
   at NUnit.Framework.Internal.Execution.ParallelWorkItemDispatcher.Execute(WorkItem work) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\ParallelWorkItemDispatcher.cs:line 155
   at NUnit.Framework.Internal.Execution.ParallelWorkItemDispatcher.Dispatch(WorkItem work) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\ParallelWorkItemDispatcher.cs:line 145
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs:line 237
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs:line 117
   at NUnit.Framework.Internal.Execution.WorkItem.RunTest() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 360
   at NUnit.Framework.Internal.Execution.WorkItem.Execute() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 260
   at NUnit.Framework.Internal.Execution.TestWorker.TestWorkerThreadProc() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\TestWorker.cs:line 111
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.runTryCode(Object userData)
   at System.Runtime.CompilerServices.RuntimeHelpers.ExecuteCodeWithGuaranteedCleanup(TryCode code, CleanupCode backoutCode, Object userData)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```

Could we add some filtering? I.e. remove everything before the Assert method and after 'NUnit.Framework.Assert.Multiple'?